src/serializationStrategies/XpathRangeSelector/xpath.js

```diff
@@ -1,31 +1,140 @@
+  if (isText(focus) && isTextHighlight(element)) {
+    while(element && (isText(element) || isTextHighlight(element))) {
+      offset += element.textContent.length;
+      element = element.previousSibling;
+    }
+  // for element targets, highlight children might be artifically
+  // inflating the range offset, fix.
+  } else if (isElement(focus)) {
+    offset -= Array.prototype.filter.call(Array.prototype.slice.call(focus.childNodes, 0, offset), isHighlight).length;
+  }
+
+  while(focus !== reference) {
+    pos = 1;
+
+    while(element) {
+      // highlights in text change the number of nodes in the nodelist,
+      // compensate by gobbling adjacent highlights and text
+      if (isText(focus) && isTextHighlight(element)) {
+        while(element && (isText(element) || isTextHighlight(element))) {
+          element = element.previousSibling;
+        }
+      }
+      if (element) {
+        if ([3, 1].indexOf(element.nodeType) !== -1 && !isHighlight(element) && element.nodeName === focus.nodeName) { // If it is ELEMENT_NODE or TEXT_NODE of the same name
+          pos += 1;
+        }
+
+        element = element.previousSibling;
       }
-      tempitem2 = tempitem2.previousSibling;
     }
 
-    if (el.nodeType === 3 /* #text */) {
+    if (focus.nodeType === 3 /* #text */) {
       xpath = "text()["+pos+']'+'/'+xpath;
     } else {
-      xpath = "*[name()='"+el.nodeName.toLowerCase()+"']["+pos+']'+'/'+xpath;
+      xpath = "*[name()='"+focus.nodeName.toLowerCase()+"']["+pos+']'+'/'+xpath;
     }
 
-    el = el.parentNode;
+    focus = focus.parentNode;
+    element = focus.previousSibling;
+
   }
+
   xpath = './' + xpath;
   xpath = xpath.replace(/\/$/, '');
-  return xpath;
+  return [xpath, offset];
+}
+
+export function getFirstByXPath(path, offset, referenceElement) {
+  const parts = path.split('/');
+
+  let node = referenceElement;
+  let part = parts.shift();
+
+  while (node && part) {
+    node = followPart(node, part);
+    part = parts.shift();
+  }
+
+  const highlightOrText = node => isHighlight(node) || isText(node);
+
+  // highligts split up text nodes that should be treated as one, iterate through
+  // until we find the text node that the offset specifies, modifying the offset
+  // as we go. prefer leaving highlights if we have the option to deal with
+  // adjacent highlights.
+  while ((isHighlight(node) && offset >= node.textContent.length) || (isText(node) && offset > node.textContent.length)) {
+    offset -= node.textContent.length;
+    node = highlightOrText(node.nextSibling) ? node.nextSibling : null;
+  }
+
+  if (node && isHighlight(node)) {
+    node = null;
+  }
+
+  return [node, offset];
 }
 
-export function getFirstByXPath(path, referenceElement) {
-  return document.evaluate(path, referenceElement).iterateNext();
+function followPart(node, part) {
+  const self = /^\.$/;
+  const textTest = /^text\(\)\[(\d+)\]$/;
+  const elementTest = /\*\[name\(\)='(.+)'\]\[(\d)\]/;
+
```

Comment:
my personal preference for constants like this is to put them at the top of the file in all caps, and prefix a `IS` if they're a test like: 

```js
IS_SELF =  /^\.$/;
IS_TEXT =/^text\(\)\[(\d+)\]$/;
IS_ELEMENT = /\*\[name\(\)='(.+)'\]\[(\d)\]/;
```

that way they're not re-declared every time the function is called and are immediately obvious that they're a both a test and a const.

Also the `textTest` reads kinda funny when `.test` is called on it bc it's then: `textTest.test`

Reply:
updated with more verbose names